consensus/polybft/consensus_runtime.go

```diff
@@ -167,15 +157,41 @@ func newConsensusRuntime(log hcf.Logger, config *runtimeConfig) (*consensusRunti
 			log.Named("checkpoint_manager"))
 	}
 
+	// we need to call restart epoch on runtime to initialize epoch state
+	runtime.epoch, err = runtime.restartEpoch(runtime.lastBuiltBlock)
+	if err != nil {
+		return nil, fmt.Errorf("consensus runtime creation - restart epoch failed: %w", err)
+	}
+
 	return runtime, nil
 }
 
-// getEpoch returns current epochMetadata in a thread-safe manner.
-func (c *consensusRuntime) getEpoch() *epochMetadata {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
+// initStateSyncManager initializes state sync manager
+// if bridge is not enabled, then a dummy state sync manager will be used
+func (c *consensusRuntime) initStateSyncManager(log hcf.Logger) error {
+	if c.IsBridgeEnabled() {
+		stateSyncManager, err := NewStateSyncManager(
+			log,
+			c.config.State,
+			&stateSyncConfig{
+				key:             c.config.Key,
+				stateSenderAddr: c.config.PolyBFTConfig.Bridge.BridgeAddr,
+				jsonrpcAddr:     c.config.PolyBFTConfig.Bridge.JSONRPCEndpoint,
+				dataDir:         c.config.DataDir,
+				topic:           c.config.bridgeTopic,
+			},
+		)
 
-	return c.epoch
+		if err != nil {
+			return err
+		}
+
+		c.stateSyncManager = stateSyncManager
+	} else {
+		c.stateSyncManager = &dummyStateSyncManager{}
```

Comment:
If the `stateSyncManager` is not enabled because the bridge is not enabled, it should never be called. I understand this is a safety measure in case it is ever called. But, I would prefer if we ensure on tests that never happens instead of having extra code to ensure that.

Note that once we have hooks (if we ever do it though) we do not need this check anymore.

Reply:
While I do understand your point, I must agree with Stefan here. `Nullable` pattern is a good practice, since it reduces the unnecessary checks and conditions in code, makes testing easier, and it doesn't really add any more complexity (since that `nil` or `dummy` implementation doesn't do anything). Plus, as you said, it keeps as out of harms way of nil exceptions. To me it makes sense to have it here, and if we ever implement that hook system, then I would agree, we can remove this `nil` implementation since we would not need it.

Reply:
The other option would be to make `StateSyncManager` aware that he is not active and do nothing when `PostEpoch` and `PostBlock` get called.

Reply:
Yes, but that means we will just be moving those checks to the `stateSyncManager` :confused: . Those hooks are looking better by the minute :smile: .